src/main/java/com/google/api/gax/grpc/ApiCallSettings.java

```diff
@@ -44,44 +44,72 @@
  */
 public class ApiCallSettings {
 
-  private Set<Status.Code> retryableCodes = new HashSet<>();
-  private RetryParams retryParams = null;
+  private final ImmutableSet<Status.Code> retryableCodes;
+  private final RetrySettings retrySettings;
 
-  /**
-   * Sets the retryable codes.
-   */
-  public ApiCallSettings setRetryableCodes(Set<Status.Code> retryableCodes) {
-    this.retryableCodes = retryableCodes;
-    return this;
+  public ImmutableSet<Status.Code> getRetryableCodes() {
+    return retryableCodes;
   }
 
-  /**
-   * Sets the retryable codes.
-   */
-  public ApiCallSettings setRetryableCodes(Status.Code... codes) {
-    this.retryableCodes = Sets.newHashSet(codes);
-    return this;
+  public RetrySettings getRetrySettings() {
+    return retrySettings;
   }
 
-  /**
-   * Gets the retryable codes.
-   */
-  public Set<Status.Code> getRetryableCodes() {
-    return retryableCodes;
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public Builder toBuilder() {
+    return new Builder(this);
   }
 
-  /**
-   * Sets the retry params.
-   */
-  public ApiCallSettings setRetryParams(RetryParams retryParams) {
-    this.retryParams = retryParams;
-    return this;
+  protected ApiCallSettings(ImmutableSet<Status.Code> retryableCodes,
+                            RetrySettings retrySettings) {
+    this.retryableCodes = ImmutableSet.<Status.Code>copyOf(retryableCodes);
+    this.retrySettings = retrySettings;
   }
 
-  /**
-   * Returns the retry params.
-   */
-  public RetryParams getRetryParams() {
-    return retryParams;
+  public static class Builder {
+
+    private Set<Status.Code> retryableCodes;
+    private RetrySettings.Builder retrySettingsBuilder;
+
+    public Builder() {
+      retryableCodes = Sets.newHashSet();
+      retrySettingsBuilder = RetrySettings.newBuilder();
+    }
+
+    public Builder(ApiCallSettings apiCallSettings) {
+      this.retryableCodes = Sets.newHashSet(apiCallSettings.retryableCodes);
+      this.retrySettingsBuilder = apiCallSettings.getRetrySettings().toBuilder();
+    }
+
+    public Builder setRetryableCodes(Set<Status.Code> retryableCodes) {
+      this.retryableCodes = Sets.newHashSet(retryableCodes);
+      return this;
+    }
+
+    public Builder setRetryableCodes(Status.Code... codes) {
+      this.setRetryableCodes(Sets.newHashSet(codes));
+      return this;
+    }
+
+    public Builder setRetrySettingsBuilder(RetrySettings.Builder retrySettingsBuilder) {
+      this.retrySettingsBuilder = retrySettingsBuilder.build().toBuilder();
```

Comment:
We don't necessarily have to do a defensive copy in this particular spot. Also, this forces the retry settings to all be set.